Source/Core/ClippingPlaneCollection.js

```diff
@@ -237,50 +326,149 @@ define([
+        result.y = octEncodeScratch.x;
+        result.z = octEncodeScratch.y * rightShift;
+        result.w = octEncodeScratch.y;
+        return result;
+    }
 
-        var planes = this._planes;
+    var oct32EncodeScratch = new Cartesian4();
+    function packAndReturnChanged(clippingPlaneCollection) {
+        var previousUint32View = clippingPlaneCollection._previousUint32View;
+        var currentUint32View = clippingPlaneCollection._currentUint32View;
+        var textureBytes = clippingPlaneCollection._textureBytes;
+        var planes = clippingPlaneCollection._planes;
         var length = planes.length;
+        var unchanged = true;
+        var byteIndex, uint32Index;
 
-        var index = 0;
-        if (!defined(array)) {
-            array = new Array(length);
-        } else {
-            index = array.length;
-            array.length = length;
-        }
+        var lengthRangeUnion = clippingPlaneCollection._lengthRangeUnion;
+        lengthRangeUnion.x = length;
+        lengthRangeUnion.w = clippingPlaneCollection._unionClippingRegions ? 1.0 : 0.0;
 
+        // Pack all plane normals and get min/max of all distances.
+        // Check each normal for changes.
+        var distanceMin = Number.POSITIVE_INFINITY;
+        var distanceMax = Number.NEGATIVE_INFINITY;
         var i;
-        for (i = index; i < length; ++i) {
-            array[i] = new Cartesian4();
-        }
-
-        var transform = Matrix4.multiply(viewMatrix, this.modelMatrix, scratchMatrix);
-
         for (i = 0; i < length; ++i) {
             var plane = planes[i];
-            var packedPlane = array[i];
 
-            Plane.transform(plane, transform, scratchPlane);
+            byteIndex = i * 8; // each plane is 4 bytes normal, 4 bytes distance
+
+            var oct32Normal = oct32EncodeNormal(plane.normal, oct32EncodeScratch);
+            textureBytes[byteIndex] = oct32Normal.x;
+            textureBytes[byteIndex + 1] = oct32Normal.y;
+            textureBytes[byteIndex + 2] = oct32Normal.z;
+            textureBytes[byteIndex + 3] = oct32Normal.w;
+
+            uint32Index = i + i;
+            unchanged = unchanged && previousUint32View[uint32Index] === currentUint32View[uint32Index];
+            if (!unchanged) {
+                previousUint32View[uint32Index] = currentUint32View[uint32Index];
+            }
 
-            Cartesian3.clone(scratchPlane.normal, packedPlane);
-            packedPlane.w = scratchPlane.distance;
+            var distance = plane.distance;
+            distanceMin = Math.min(distance, distanceMin);
+            distanceMax = Math.max(distance, distanceMax);
         }
 
-        return array;
+         // Expand distance range a little bit to prevent packing 1s
+         distanceMax += (distanceMax - distanceMin) * CesiumMath.EPSILON3;
+
+         // Normalize all the distances to the range and record them in the typed array.
+         // Check each distance for changes.
```

Comment:
I think in those cases precision issues are more likely to manifest as just not being able to place the plane at centimeter-scale "correctness" as opposed to jitter.

That being said, having read the [vector paper's](http://jcgt.org/published/0003/02/01/) intro on the philosophy behind finite bit-patterns representing finite numbers, it's definitely wasteful to use this "normalizing" method for compressing arbitrary floats :(

Perhaps we should look into packing/unpacking the actual bit pattern instead of a normalization?

Reply:
> Perhaps we should look into packing/unpacking the actual bit pattern instead of a normalization?

Okay I think actually doing that is insane which is why there's so many approximations. What I meant was "maybe we should look into other approximations." Or just use float textures when available, as recommended by @pjcozzi, but that also needs a little more thought about precisions (texture access always returns [0, 1] right?) and I think that adds enough complexity to warrant a separate PR.

Reply:
> texture access always returns [0, 1] right?

Apparently not :D
After a small discussion offline, I'll take a look at just enabling floating point textures as the default path, in which case the RGBA code will become a slightly complicated fallback.